File: src/net/liplum/lib/shaders/ProgressShader.java
package net.liplum.lib.shaders;

import arc.Core;
import arc.graphics.gl.Shader;
import arc.util.Time;
import mindustry.Vars;
import net.liplum.R;

import static mindustry.Vars.renderer;
import static mindustry.graphics.Shaders.getShaderFi;

public class ProgressShader extends Shader implements IReusable {
    public float progress = 0f;

    public ProgressShader(String fragName) {
        super(getShaderFi("default.vert"),
                Vars.tree.get(R.S.GenFrag(fragName)));
    }

    @Override
    public void apply() {
        setUniformf("u_time", Time.time);
        setUniformf("u_progress", progress);
        setUniformf("u_resolution",
                Core.graphics.getWidth(),
                Core.graphics.getHeight()
        );
        setUniformf("u_offset",
                Core.camera.position.x,
                Core.camera.position.y
        );
        renderer.effectBuffer.getTexture().bind(0);
    }

    @Override
    public void reset() {
        this.progress = 0f;
    }
}

File: main/src/net/liplum/shaders/ProgressShader.kt
package net.liplum.shaders

import arc.Core
import arc.files.Fi
import arc.util.Time
import mindustry.Vars
import net.liplum.common.shader.ShaderBase

class ProgressShader(
    vert: Fi, frag: Fi,
) : ShaderBase(vert, frag) {
    var progress = 0f
    override fun apply() {
        setUniformf("u_time", Time.time)
        setUniformf("u_progress", progress)
        setUniformf(
            "u_resolution",
            Core.graphics.width.toFloat(),
            Core.graphics.height
                .toFloat()
        )
        setUniformf(
            "u_offset",
            Core.camera.position.x,
            Core.camera.position.y
        )
        Vars.renderer.effectBuffer.texture.bind(0)
    }

    override fun reset() {
        super.reset()
        progress = 0f
    }
}